cpp/src/arrow/ipc/writer.cc

```diff
@@ -409,7 +409,7 @@ class RecordBatchSerializer : public ArrayVisitor {
 
       // Allocate an array of child offsets. Set all to -1 to indicate that we
       // haven't observed a first occurrence of a particular child yet
-      std::vector<int32_t> child_offsets(max_code + 1);
+      std::vector<int32_t> child_offsets(max_code + 1, -1);
```

Comment:
Note these are unrelated logic fixes I spotted while reading this code (this is the path for dense unions).
If desired I can separate those changes into a separate PR. In any case, it needs validating by a reviewer.